lib/spack/spack/spec_list.py

```diff
@@ -0,0 +1,167 @@
+                if isinstance(item, dict):  # matrix of specs
+                    excludes = item.get('exclude', [])
+                    for combo in itertools.product(*(item['matrix'])):
+                        # Test against the excludes using a single spec
+                        ordered_combo = sorted(combo, key=spec_ordering_key)
+                        test_spec = Spec(' '.join(ordered_combo))
+                        if any(test_spec.satisfies(x) for x in excludes):
+                            continue
+
+                        # Add as list of constraints
+                        constraints.append([Spec(x) for x in ordered_combo])
+                else:  # individual spec
+                    constraints.append([Spec(item)])
+            self._constraints = constraints
+
+        return self._constraints
+
+    @property
+    def specs(self):
+        if self._specs is None:
+            specs = []
+            # This could be slightly faster done directly from yaml_list,
+            # but this way is easier to maintain.
+            for constraint_list in self.specs_as_constraints:
+                spec = constraint_list[0].copy()
+                for const in constraint_list[1:]:
+                    spec.constrain(const)
+                specs.append(spec)
+            self._specs = specs
+
+        return self._specs
+
+    def add(self, spec):
+        self.yaml_list.append(str(spec))
+
+        # expanded list can be updated without invalidation
+        if self._expanded_list is not None:
+            self._expanded_list.append(str(spec))
+
+        # Invalidate cache variables when we change the list
+        self._constraints = None
+        self._specs = None
+
+    def remove(self, spec):
+        # Get spec to remove from list
+        remove = [s for s in self.yaml_list
+                  if (isinstance(s, string_types) and not s.startswith('$'))
+                  and Spec(s) == Spec(spec)]
+        if not remove:
+            msg = 'Cannot remove %s from SpecList %s\n' % (spec, self.name)
+            msg += 'Either %s is not in %s or %s is ' % (spec, self.name, spec)
+            msg += 'expanded from a matrix and cannot be removed directly.'
+            raise SpecListError(msg)
+        assert len(remove) == 1
+        self.yaml_list.remove(remove[0])
+
+        # invalidate cache variables when we change the list
+        self._expanded_list = None
+        self._constraints = None
+        self._specs = None
+
+    def extend(self, other, copy_reference=True):
+        self.yaml_list.extend(other.yaml_list)
+        self._expanded_list = None
+        self._constraints = None
+        self._specs = None
+
+        if copy_reference:
+            self._reference = other._reference
+
+    def update_reference(self, reference):
+        self._reference = reference
+        self._expanded_list = None
+        self._constraints = None
+        self._specs = None
+
+    def _expand_references(self, yaml):
```

Comment:
I'd like to see `yaml` eventually (not necessarily now) become some data structure rather than raw YAML.  Like maybe it should take a `dict` of `SpecLists`.  It is recursive, right?

This keeps the details of the YAML format (where, for aesthetic reasons, we distinguish between a single string element and a single-element `dict` containing a `list`) are kept separate from the `SpecList` logic.  Doing this also makes the function more testable and the `SpecList` class more reusable.